File: 0039-combination-sum/0039-combination-sum.java
class Solution {
    public static List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<List<Integer>> list = new ArrayList<>();
        fn(0, list,new ArrayList<>(), candidates, target);
        System.out.println(list);
        return list;
    }

     public static void fn(int i,List<List<Integer>> list, List<Integer> l, int[] arr,int target){
       if(i>=arr.length){
        if(target==0) list.add(new ArrayList<>(l));
        return;
       }

       if(arr[i]<=target){
            l.add(arr[i]);
            fn(i,list,l,arr,target-arr[i]);
            l.remove(l.size()-1);
       }

       fn(i+1,list,l,arr,target);
    }
}

File: 0039-combination-sum/0039-combination-sum.cpp
class Solution {
public:
    vector<vector<int>> res;
    
    void dfs(vector<int>& candidates, int target, int idx, vector<int>& path) {
        if (target == 0) {
            res.push_back(path);
            return;
        }
        if (idx >= candidates.size() || target < 0) return;
        
        path.push_back(candidates[idx]);
        dfs(candidates, target - candidates[idx], idx, path);
        path.pop_back();
        
        dfs(candidates, target, idx + 1, path);
    }
    
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        vector<int> path;
        dfs(candidates, target, 0, path);
        return res;
    }
};
